lib/Surface/USB/Shuttle.js

```diff
@@ -0,0 +1,202 @@
+	/*	totalCols: 7,
+	totalRows: 3,
+
+	encoders: [0, 7],
+	buttons: [14, 6, 13, 20],*/
+}
+const shuttleProV2Info = {
+	// TODO(Peter): Improve this mapping
+	totalCols: 5,
+	totalRows: 3,
+
+	encoders: [16],
+	buttons: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
+}
+
+function buttonToIndex(modelInfo, info) {
+	const index = modelInfo.buttons[info - 1]
+	if (index !== undefined) {
+		return index
+	}
+
+	return undefined
+}
+
+function rotaryToButtonIndex(modelInfo, info) {
+	const index = modelInfo.encoders[info - 1]
+	if (index !== undefined) {
+		return index
+	}
+
+	return undefined
+}
+
+class SurfaceUSBShuttle {
+	constructor(ipcWrapper, devicepath, shuttle, modelInfo, deviceInfo) {
+		this.ipcWrapper = ipcWrapper
+		this.shuttle = shuttle
+		this.deviceInfo = deviceInfo
+		this.modelInfo = modelInfo
+
+		this.config = {
+			// No config currently present
+		}
+
+		this.ipcWrapper.log('debug', `Adding Contour Shuttle USB device ${devicepath}`)
+
+		this.info = {
+			type: `Shuttle ${this.deviceInfo.name}`,
+			devicepath: devicepath,
+			configFields: [],
+			keysPerRow: this.modelInfo.totalCols,
+			keysTotal: this.modelInfo.totalCols * this.modelInfo.totalRows,
+			deviceId: `shuttle:${this.deviceInfo.id}`,
+		}
+
+		this.shuttle.on('error', (error) => {
+			console.error(error)
+			this.ipcWrapper.remove()
+		})
+
+		this.shuttle.on('buttondown', (info) => {
+			console.log('button down', info)
+			const keyIndex = buttonToIndex(this.modelInfo, info) ?? rotaryToButtonIndex(this.modelInfo, info)
+			console.log('button down key index', keyIndex)
+			if (keyIndex === undefined) {
+				return
+			}
+
+			this.ipcWrapper.click(keyIndex, true)
+		})
+
+		this.shuttle.on('buttonup', (info) => {
+			console.log('button up', info)
+			const keyIndex = buttonToIndex(this.modelInfo, info) ?? rotaryToButtonIndex(this.modelInfo, info)
+			console.log('button up key index', keyIndex)
+			if (keyIndex === undefined) {
+				return
+			}
+
+			this.ipcWrapper.click(keyIndex, false)
+		})
+		this.shuttle.on('jog', (info, delta) => {
+			const keyIndex = rotaryToButtonIndex(this.modelInfo, info)
+			if (keyIndex === undefined) {
+				return
+			}
+
+			this.ipcWrapper.rotate(keyIndex, delta == 1)
+		})
+
+		this.shuttle.on('disconnect', (error) => {
+			console.error(error)
+			this.ipcWrapper.remove()
+		})
+	}
+
+	async #init() {
+		this.ipcWrapper.log('debug', `Contour Shuttle ${this.deviceInfo.name} detected`)
+
+		// Make sure the first clear happens properly
+		// TODO(Peter): Is this required?
+		//await this.loupedeck.blankDevice(true, true)
+	}
+
+	static async create(ipcWrapper, devicepath) {
+		const shuttle = shuttleControlUSB
+		shuttle.start()
+		try {
+			let deviceInfo = null
+			let info = null
+			console.log('Device count', shuttle.getDeviceList().length)
+			if (shuttle.getDeviceList().length >= 1) {
+				// TODO(Peter): Fix this to fetch the correct device
+				deviceInfo = shuttle.getDeviceList()[0]
```

Comment:
This won't work on macos. macos is different to windows and linux in that HID devices can only be opened in one place at a time.
Even doing two opens of a device from the same process/thread does not work.
![Screenshot from 2023-05-07 14-22-50](https://user-images.githubusercontent.com/1327476/236680527-60bf3313-da9a-4a8f-b4a1-d64301922032.png)


In the current implementation here, when two devices are connected a different child_process will be spawned for each. Both of those will be running the usb watcher and both will attempt to open the hid devices. Only one will succeed with the opening, and there is no guarantee it will be the one we want it to be.

-----

This either needs to not run the usb hotplug monitoring in each child_process and instead explicitly open a specified path, or it should be running all of the shuttles in a single child_process, and ignore them in the main companion hotplug detection

Reply:
> This won't work on macos. macos is different to windows and linux in that HID devices can only be opened in one place at a time. Even doing two opens of a device from the same process/thread does not work.

Out of interest, is that a core Mac limitation, rather than a node-hid or hidapi specific?

> In the current implementation here, when two devices are connected a different child_process will be spawned for each. Both of those will be running the usb watcher and both will attempt to open the hid devices. Only one will succeed with the opening, and there is no guarantee it will be the one we want it to be.

I assume that just calling `node-hid.devices()` is enough to cause this issue?

> This either needs to not run the usb hotplug monitoring in each child_process and instead explicitly open a specified path, or it should be running all of the shuttles in a single child_process, and ignore them in the main companion hotplug detection

I'm assuming from https://github.com/bitfocus/companion/issues/2430#issuecomment-1528896873 that your preference is the former? I'll see if I can update the package to do that if you're happy with that too @hopejr ?

Reply:
> Out of interest, is that a core Mac limitation, rather than a node-hid or hidapi specific?

Oh, maybe it can be turned off, but it isnt exposed in `node-hid` currently. I suppose looking into that is an option, but it still feels rather wasteful to be opening them all multiple times.

> I assume that just calling node-hid.devices() is enough to cause this issue?

Its the `new hid.HID(device.path)` call that will be problematic. doing a `hid.devices()` is perfectly fine (in this context)

> I'm assuming from https://github.com/bitfocus/companion/issues/2430#issuecomment-1528896873 that your preference is the former?

Yeah, that would be prefered. It will avoid adding a lot of complexity to companion to handle this, at what I hope is a small cost to the library.  
In develop (3.x), this child_process stuff is being removed, so it might not be as bad, but we still have the same system for scanning for devices.

Reply:
@peternewman if you can come up with a way that's not going to break backwards compatibility, then I have no problem with it. Though does getRawHidDevice() not do what you need? We could always change it to accept path or id (the find function can do ele => ele.id === id || ele.path === id).

Reply:
Thanks @hopejr . 

How does this look?:
https://github.com/hopejr/ShuttleControlUSB/pull/8/files

Unfortunately `getRawHidDevice()` on it's own is no good, because all the `new hid.HID(device.path)` @Julusian refers to still get called.

Reply:
We're now not scanning with @hopejr 's library so hopefully this is sorted now.